### PR TITLE
fix(shared): truncate() no longer splits surrogate pairs (#16172)

### LINKED ISSUE #16172: bug: truncate() splits Unicode surrogate pairs, producing a replacement character
### Describe the bug

`truncate()` in `packages/shared/src/utils/stringChecks.ts` cuts strings by UTF-16 code units via `String.prototype.substring`, with no awareness of surrogate pairs:

```ts
export function truncate(str: string, n = 16) {
  if (str.length > n) {
    return str.substring(0, n) + "...";
  }
  return str;
}
```

Any character outside the Basic Multilingual Plane — emoji (🚀, 🎉), many CJK extension-B ideographs, mathematical alphanumerics (𝟙), etc. — is stored as two UTF-16 code units (a surrogate pair). When the truncation boundary `n` falls between the two halves of a pair, `substring(0, n)` keeps only the high surrogate. The result is a lone, unpaired surrogate that renders as the Unicode replacement character `�`.

This is user-visible wherever `truncate()` is applied to strings that can contain such characters, including:

- `web/src/features/dashboard/components/ModelCostTable.tsx:90` — `truncate(item.providedModelName as string, 30)` (model names are user-controlled and frequently contain emoji)
- `web/src/features/dashboard/components/LatencyTables.tsx:124,127` — `truncate(item.name as string)` (trace/observation names, default n = 16)
- `packages/shared/src/in-app-agent/server/persistence.ts:1562` — `truncate(normalizedText, 600)`

Whether a given value breaks depends on the offset of the astral character relative to `n`, so it fails intermittently and looks like data corruption to the user.

**Expected:** truncation never splits a code point; output is always well-formed text (worst case, one character shorter).
**Actual:** output can end in a lone high surrogate that displays as `�`.

### Steps to reproduce

Minimal, deterministic reproduction of the helper (Node):

```js
function truncate(str, n = 16) {
  return str.length > n ? str.substring(0, n) + "..." : str;
}

// 'a' shifts the emoji pairs onto an odd boundary so the default n=16 cut lands mid-pair
console.log(JSON.stringify(truncate("a" + "🎉".repeat(12))));
// -> "a🎉🎉🎉🎉🎉🎉🎉\ud83c..."  (lone high surrogate -> renders as ...�...)

// realistic model-name case, n=30
console.log(truncate("gpt-4o-🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀-preview", 30));
// -> "gpt-4o-🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀�..."
```

User-facing path: open **Dashboard → Model Costs** for a project with traces on a model whose display name contains an emoji positioned so the name exceeds 30 code units at an odd boundary (e.g. `gpt-4o-🚀-experimental-long-name`). The truncated cell shows a trailing `�`.

### Langfuse Cloud or self-hosted?

Self-hosted — found by source review of `main` @ ab58010; same shared code applies to Langfuse Cloud.

### Additional information

Root cause is code-unit vs. code-point truncation. Suggested fix — iterate by code point so a pair is never split:

```ts
export function truncate(str: string, n = 16) {
  const codePoints = Array.from(str); // splits on code points, keeps pairs intact
  if (codePoints.length > n) {
    return codePoints.slice(0, n).join("") + "...";
  }
  return str;
}
```

This changes `n` to mean code points rather than UTF-16 code units, which is almost certainly the intended semantics for a display-truncation helper. For fully grapheme-safe truncation (emoji ZWJ sequences, skin-tone modifiers), `Intl.Segmenter` with `granularity: "grapheme"` is more thorough, at some CPU cost. A unit test covering an astral character straddling the boundary would lock this in.

### Are you interested in contributing a fix?

Yes.

### PR DESCRIPTION
**Pull request by greekera1000**
## What does this PR do?

`truncate()` in `packages/shared/src/utils/stringChecks.ts` sliced strings with `str.substring(0, n)`, which counts UTF-16 code units. When `n` landed between the two halves of a surrogate pair (any astral character — emoji, CJK extension-B, mathematical alphanumerics), it kept only the high surrogate, leaving a lone surrogate that renders as the replacement character `�`.

This surfaces wherever the shared `truncate()` is used on user-controlled text, e.g. `ModelCostTable` and `LatencyTables` (model/trace names) and the in-app-agent path.

The fix iterates by code point via `Array.from(str)` before slicing, so a pair is never split. `n` now counts code points rather than UTF-16 code units, which matches the intended semantics of a display-truncation helper.

Adds `stringChecks.test.ts` (the module had no tests) covering the surrogate-boundary cases plus BMP/ellipsis controls. All 7 pass locally via vitest; the surrogate cases fail against the old implementation, confirming they exercise the bug.

Fixes #16172

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Chore (tooling, dependencies, CI, workflows, repo upkeep, or other maintenance work)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Refactor (restructures existing code without changing behavior, e.g. simplify logic, split modules, reduce duplication)
- [ ] This change requires a documentation update

## Mandatory Tasks

- [x] Make sure you have self-reviewed the code. A decent size PR without self-review might be rejected.

<details open><summary><h3>Greptile Summary</h3></summary>

The PR updates the shared `truncate()` helper to count Unicode code points and avoid splitting UTF-16 surrogate pairs.
- Converts strings to code-point arrays before checking and slicing their length.
- Adds focused tests for astral characters, BMP text, ellipsis behavior, and unchanged strings.
</details>

<details><summary><h3>Confidence Score: 5/5</h3></summary>

The PR appears safe to merge with no actionable defects identified.

The implementation preserves existing behavior for current positive integer limits while preventing truncation from producing lone UTF-16 surrogates, and the added tests cover the changed behavior.
</details>

<sub>Reviews (1): Last reviewed commit: ["fix(shared): truncate() no longer splits..."](https://github.com/langfuse/langfuse/commit/f4670f71c4b26927769b5261f96896e5d55fa155) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=54318841)</sub>

### CODE REVIEW

**Comment by nimarb** [comment]
@claude review

File: packages/shared/src/utils/stringChecks.test.ts [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+  it("appends an ellipsis when over the limit", () => {
+    expect(truncate("this is definitely longer than sixteen", 16)).toBe(
+      "this is definite...",
+    );
+  });
+
+  it("counts code points, not UTF-16 code units, for the limit", () => {
+    expect(truncate("🎉".repeat(8), 16)).toBe("🎉".repeat(8));
+  });
```
[the remark this answers is not in the document]
Comment (greekera1000) [reply packages/shared/src/utils/stringChecks.test.ts]: Good catch, thanks — you're right that `.repeat(8)` at n=16 sits exactly on the code-unit boundary, so it passed against the old implementation too. Changed it to `.repeat(9)` (commit cc5a7ed) so the two implementations actually diverge. The surrogate-boundary tests at lines 35–40 and 49–54 remain the ones that genuinely fail pre-fix.

**Comment by nimarb** [comment]
@claude review
